rtrlib/lib/vector.c

```diff
@@ -0,0 +1,80 @@
+
+
+#include "vector.h"
+
+int create_vector(struct lrtr_vector **vector_pointer)
+{
+	const size_t DEFAULT_INITIAL_SIZE = 100;
+
+	// allocation the chunk of memory
+	uint32_t *data_field = (uint32_t *)malloc(sizeof(uint32_t) * DEFAULT_INITIAL_SIZE);
+
+	// malloc failed so returning an error
+	if (data_field == NULL) {
+		return -1;
+	}
+
+	struct lrtr_vector *vector = (struct lrtr_vector *)malloc(sizeof(struct lrtr_vector));
+
+	if (vector == NULL) {
+		return -1;
+	}
+
+	vector->capacity = DEFAULT_INITIAL_SIZE;
+	vector->size = 0;
+	vector->data = data_field;
+
+	*vector_pointer = vector;
+
+	return 0;
+}
+
+int delete_vector(struct lrtr_vector *vector)
+{
+	// checking if one of the two pointers is potentially invalid
+	if (vector == NULL || vector->data == NULL) {
+		return -1;
+	}
+
+	// freeing the data
+	free(vector->data);
+	// freeing the object itself
+	free(vector);
+
+	return 0;
+}
+
+int reallocate_vector(struct lrtr_vector *vector)
+{
+	// by how many more elements this vector can hold after the relocation
+	const size_t SIZE_INCREASE = 100;
+
+	// allocation the new chunk of memory
+	uint32_t *new_data_field = (uint32_t *)malloc(sizeof(uint32_t) * (vector->capacity + SIZE_INCREASE));
+
+	// malloc failed so returning an error
+	if (new_data_field == NULL) {
+		return -1;
+	}
+
+	// copying the data from the old location to the new one
+	memcpy(new_data_field, vector->data, vector->capacity);
+	vector->capacity += SIZE_INCREASE;
+
+	return 0;
+}
+
+int add_element_vector(struct lrtr_vector *vector, uint32_t value)
```

Comment:
suggestion (optional): add a function to add an array of elements, not requiring the size check for each inserted element.